github_poster/loader/todoist_loader.py

```diff
@@ -0,0 +1,142 @@
+        res = requests.post(url=url, data=postdata, headers=headers)
+        resposn = res.json()
+        return resposn
+
+    # call with re-try since todoist api some time get 502
+    def response_with_retry(self, url, postdata, times):
+        # time.sleep(1)
+        try:
+            return self.response(url, postdata)
+        except Exception as e:
+            if times >= self.MAXIMAL_RETRY:
+                print(
+                    f">> Exceed maximal retry {self.MAXIMAL_RETRY}, Raise exception..."
+                )
+                raise (e)  # will stop the program without further handling
+            else:
+                times += 1
+                print(f">> Exception, Retry {times} begins...")
+                return self.response_with_retry(url, postdata, times)
+
+    # call todoist dev api to get activity of completed tasks
+    # refer https://developer.todoist.com/sync/v9/#activity
+    def todoist_completed_activity(self, page, limit, offset):
+        data = {
+            "event_type": "completed",
+            "page": page,
+            "limit": limit,
+            "offset": offset,
+        }
+        url = "https://api.todoist.com/sync/v9/activity/get"
+        re = self.response_with_retry(url, data, 1)
+        return re
+
+    # json expect to be list of events format
+    # with event_date, event_type, id
+    def normalize_df(self, jsondata):
+        if jsondata["count"] == 0:
+            return pd.DataFrame(columns=["event_date", "event_type", "id"])
+        df = pd.json_normalize(jsondata["events"])
+        df = df[["event_date", "event_type", "id"]]
+        df["event_date"] = df["event_date"].str.slice(0, 10)
+        return df
+
+    def count_to_dict(self, df):
+        return df.groupby(["event_date"])["event_date"].count().to_dict()
+
+    # refer https://developer.todoist.com/sync/v9/#activity
+    # todoist api only allows you to get activity data by page from current day
+    # we will have to calculate the pages based on from and to year then manipulate the dict data
+    def get_api_data(self):
+        # init critical dates
+        today = datetime.datetime.today().strftime("%Y-%m-%d")
+        current_year = datetime.datetime.now().year
+        # 52.14 weeks a year, add 53 pages per full year
+        number_of_days = datetime.date.today().timetuple().tm_yday + 365 * (
+            current_year - self.from_year
+        )
+        # current year
+        page_from = (
+            0
+            if current_year == self.to_year
+            else datetime.date.today().timetuple().tm_yday // 7
+        )
+        page_to = number_of_days // 7 + 1
+        print("Todoist API Page range ({0},{1})".format(page_from, page_to))
+        last_day_of_to_year = datetime.datetime(self.to_year, 12, 31).strftime(
+            "%Y-%m-%d"
+        )
+        first_day_of_from_year = datetime.datetime(self.from_year, 1, 1).strftime(
+            "%Y-%m-%d"
+        )
```

Comment:
use pendulum maybe easier you can check others loader

Reply:
Let me learn pendulum first...new to python and not familiar with libs like this. :)